---
title: Expected Errors
description: Explore how Effect represents and manages expected errors. Learn about creating error instances, tracking errors at the type level, and the short-circuiting behavior of Effect programs. Discover techniques to catch and recover from errors, and gain insights into error handling strategies using Effect's powerful combinators.
sidebar:
  order: 1
---

import { Aside } from "@astrojs/starlight/components"

Expected errors **are tracked at the type level** by the [Effect data type](/getting-started/the-effect-type/) in the "Error" channel:

```ts file=<rootDir>/expected-errors/program.ts

```

The type of `program` shows that it can fail with an error of type `HttpError`:

```ts "HttpError" showLineNumbers=false
const program: Effect<never, HttpError, never>
```

We use a class to represent the `HttpError` type above simply to gain access
to both the error type and a free constructor. However, you can use whatever
you like to model your error types.

It's worth noting that added a `readonly _tag` field to the `HttpError` class to serve as a discriminant for the error:

```ts {2} showLineNumbers=false
class HttpError {
  readonly _tag = "HttpError"
}
```

This will be useful later when we discuss APIs like [Effect.catchTag](#catchtag).

<Aside type="tip">
  Adding a discriminant field, such as `_tag`, can be beneficial for
  distinguishing between different types of errors during error handling. It
  also prevents TypeScript from unifying types, ensuring that each error is
  treated uniquely based on its discriminant value.
</Aside>

## Error Tracking

The following program serves as an illustration of how errors are automatically tracked for you:

<div style="display: none;">

</div>

```ts file=<rootDir>/expected-errors/error-tracking.ts

```

Effect automatically keeps track of the possible errors that can occur during the execution of the program:

```ts "HttpError | ValidationError" showLineNumbers=false
const program: Effect<string, HttpError | ValidationError, never>
```

indicating that it can potentially fail with either a `HttpError` or a `ValidationError`.

## Short-Circuiting

When working with APIs like [Effect.gen](/getting-started/using-generators/#understanding-effectgen), [Effect.map](/getting-started/using-pipelines/#map), [Effect.flatMap](/getting-started/using-pipelines/#flatmap), [Effect.andThen](/getting-started/using-pipelines/#andthen) and [Effect.all](/getting-started/using-pipelines/#all), it's important to understand how they handle errors.
These APIs are designed to **short-circuit the execution** upon encountering the **first error**.

What does this mean for you as a developer? Well, let's say you have a chain of operations or a collection of effects to be executed in sequence. If any error occurs during the execution of one of these effects, the remaining computations will be skipped, and the error will be propagated to the final result.

In simpler terms, the short-circuiting behavior ensures that if something goes wrong at any step of your program, it won't waste time executing unnecessary computations. Instead, it will immediately stop and return the error to let you know that something went wrong.

```ts file=<rootDir>/expected-errors/short-circuiting.ts

```

This code snippet demonstrates the short-circuiting behavior when an error occurs.
Each operation depends on the successful execution of the previous one.
If any error occurs, the execution is short-circuited, and the error is propagated.
In this specific example, `task3` is never executed because an error occurs in `task2`.

## Catching All Errors

### either

The `Effect.either` function transforms an `Effect<A, E, R>` into an effect that encapsulates both potential failure and success within an [Either](/data-types/either) data type:

```ts showLineNumbers=false
Effect<A, E, R> -> Effect<Either<A, E>, never, R>
```

The resulting effect cannot fail because the potential failure is now represented within the `Either`'s `Left` type.
The error type of the returned `Effect` is specified as `never`, confirming that the effect is structured to not fail.

By yielding an `Either`, we gain the ability to "pattern match" on this type to handle both failure and success cases within the generator function.

```ts collapse={3-21} file=<rootDir>/expected-errors/either-if.ts

```

We can make the code less verbose by using the `Either.match` function, which directly accepts the two callback functions for handling errors and successful values:

```ts collapse={3-21} file=<rootDir>/expected-errors/either-match.ts

```

### catchAll

The `Effect.catchAll` function allows you to catch any error that occurs in the program and provide a fallback.

```ts collapse={3-21} file=<rootDir>/expected-errors/catchAll.ts

```

We can observe that the type in the error channel has changed to `never`:

```ts showLineNumbers=false
const recovered: Effect<string, never, never>
```

indicating that all errors have been handled.

## Catching Some Errors

Suppose we want to handle a specific error, such as `HttpError`.

```ts collapse={3-21} file=<rootDir>/expected-errors/catch-some-if-http-error.ts

```

We can observe that the type in the error channel of our program has changed to only show `ValidationError`:

```ts "ValidationError" showLineNumbers=false
const recovered: Effect<string, ValidationError, never>
```

indicating that `HttpError` has been handled.

If we also want to handle `ValidationError`, we can easily add another case to our code:

```ts collapse={3-21} file=<rootDir>/expected-errors/catch-some-if-all.ts

```

We can observe that the type in the error channel has changed to `never`:

```ts showLineNumbers=false
const recovered: Effect<string, never, never>
```

indicating that all errors have been handled.

### catchSome

If we want to catch and recover from only some types of errors and effectfully attempt recovery, we can use the `Effect.catchSome` function:

```ts collapse={3-21} file=<rootDir>/expected-errors/catchSome.ts

```

In the code above, `Effect.catchSome` takes a function that examines the error (`error`) and decides whether to attempt recovery or not. If the error matches a specific condition, recovery can be attempted by returning `Option.some(effect)`. If no recovery is possible, you can simply return `Option.none()`.

It's important to note that while `Effect.catchSome` lets you catch specific errors, it **doesn't alter the error type** itself.
Therefore, the resulting effect will still have the same error type as the original effect:

```ts "HttpError | ValidationError" showLineNumbers=false
const recovered: Effect<string, HttpError | ValidationError, never>
```

### catchIf

Similar to `Effect.catchSome`, the function `Effect.catchIf` allows you to recover from specific errors based on a predicate:

```ts collapse={3-21} file=<rootDir>/expected-errors/catchIf.ts

```

It's important to note that for TypeScript versions < 5.5, while `Effect.catchIf` lets you catch specific errors, it **doesn't alter the error type** itself.
Therefore, the resulting effect will still have the same error type as the original effect:

```ts "HttpError | ValidationError" showLineNumbers=false
const recovered: Effect<string, HttpError | ValidationError, never>
```

In TypeScript versions >= 5.5, improved type narrowing causes the resulting error type to be inferred as `ValidationError`.

For TypeScript versions < 5.5, if you provide a [user-defined type guard](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) instead of a predicate, the resulting error type will be pruned, returning an `Effect<string, ValidationError, never>`:

```ts collapse={3-21} {25} file=<rootDir>/expected-errors/catchIf-guard.ts

```

### catchTag

If your program's errors are all tagged with a `_tag` field that acts as a discriminator you can use the `Effect.catchTag` function to catch and handle specific errors with precision.

```ts collapse={3-21} file=<rootDir>/expected-errors/catchTag-http-error.ts

```

In the example above, the `Effect.catchTag` function allows us to handle `HttpError` specifically.
If a `HttpError` occurs during the execution of the program, the provided error handler function will be invoked,
and the program will proceed with the recovery logic specified within the handler.

We can observe that the type in the error channel of our program has changed to only show `ValidationError`,
indicating that `HttpError` has been handled.

If we also wanted to handle `ValidationError`, we can simply add another `catchTag`:

```ts collapse={3-21} {27-29} file=<rootDir>/expected-errors/catchTag-all.ts

```

We can observe that the type in the error channel of our program has changed to `never`,
indicating that all errors have been handled.

<Aside type="caution">
  it is important to ensure that the error type used with `catchtag` has a
  `readonly _tag` discriminant field. this field is required for the matching
  and handling of specific error tags.
</Aside>

### catchTags

Instead of using the `Effect.catchTag` function multiple times to handle individual error types, we have a more convenient option called `Effect.catchTags`. With `Effect.catchTags`, we can handle multiple errors in a single block of code.

```ts collapse={3-21} file=<rootDir>/expected-errors/catchTags.ts

```

In the above example, instead of using `Effect.catchTag` multiple times to handle individual errors, we utilize the `Effect.catchTags` combinator.
This combinator takes an object where each property represents a specific error `_tag` (`"HttpError"` and `"ValidationError"` in this case),
and the corresponding value is the error handler function to be executed when that particular error occurs.

<Aside type="caution">
  It is important to ensure that all the error types used with
  `Effect.catchTags` have a `readonly _tag` discriminant field. This field is
  required for the matching and handling of specific error tags.
</Aside>
